### PR TITLE
proto 3 Rows benchmark

### PR DESCRIPTION
**Pull request by sougou**
TL;DR: The cost of manipulating and converting data structures
in go code is negligible. However the cost of decoding pointers
or multidimensional slices is very high.

This is only an experimental benchmark. I'll not be submitting
this. So, we either have to use the complicated Rows3 format,
or fix the root cause of the performance in proto3.

BenchmarkRow1-12          500000              2967 ns/op
BenchmarkMarshal1-12      200000              8587 ns/op
BenchmarkUnmarshal1-12    100000             17030 ns/op
BenchmarkRow2-12          500000              2158 ns/op
BenchmarkMarshal2-12      300000              6654 ns/op
BenchmarkUnmarshal2-12    100000             12778 ns/op
BenchmarkRow3-12         1000000              1323 ns/op
BenchmarkMarshal3-12      500000              3182 ns/op
BenchmarkUnmarshal3-12    300000              4623 ns/op

### CODE REVIEW

File: go/vt/proto/bench/bench_test.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	row[0].Values = vals
+	prow[0] = &row[0]
+	result3 := &query.Result3{Rows: prow}
+	bytes, err := proto.Marshal(result3)
+	if err != nil {
+		b.Error(err)
+		return
+	}
+	for i := 0; i < b.N; i++ {
+		res := &query.Result2{}
```
Comment (enisoc) [inline go/vt/proto/bench/bench_test.go]: Should this be Result3?

Comment (sougou) [inline go/vt/proto/bench/bench_test.go]: Oops! I'll fix & rerun.

**Comment by sougou** [comment]
I've updated the numbers after the benchmark fix. Strangely, the unmarshal for Result3 didn't get much worse. It went up by a little bit. So, the conclusions remain the same.

**Comment by erzel** [comment]
@sougou -- I think it may be a good idea to submit this. We'll have a record that we can refer to in the future if anyone wants to re-visit this.

**Comment by sougou** [comment]
I was forced to kill this because it poisons query.proto.
We'll need to rebuild using proto files that don't interfere with the production ones.
